File: VB/ViewModels/SignInViewModel.vb
Imports System.ComponentModel
Imports System.ComponentModel.DataAnnotations
Imports DevExpress.Mvvm
Imports DevExpress.Mvvm.POCO

Namespace DXHtmlMessengerSample.ViewModels
    Public Class SignInViewModel
        Implements IDocumentContent
        Public Shared Function Create(ByVal userName As String, ByVal salt As String) As SignInViewModel
            Return ViewModelSource.Create(Function() New SignInViewModel(userName, salt))
        End Function
        '
        Private ReadOnly salt As String
        Protected Sub New(ByVal userName As String, ByVal salt As String)
            Me.UserName = userName
            Me.salt = salt
        End Sub
        Public ReadOnly Property UserName() As String
        <DataType(DataType.Password)>
        Public Overridable Property Password() As String
        Private AccessTokenCore As String
        Public ReadOnly Property AccessToken() As String
            Get
                Return AccessTokenCore
            End Get
        End Property
        Public Sub ShowDialog()
            Dim flyout = Me.GetService(Of IDocumentManagerService)("Flyout")
            Dim document = flyout.CreateDocument(Me)
            document.Show()
        End Sub
        Public Sub SignInViaSocialNetwork()
            AccessTokenCore = DevExpress.DevAV.Chat.DevAVEmpployeesInMemoryServer.GetPasswordHash(String.Empty, salt)
            CloseDocument()
        End Sub
        Public Sub SignIn()
            AccessTokenCore = DevExpress.DevAV.Chat.DevAVEmpployeesInMemoryServer.GetPasswordHash(Password, salt)
            CloseDocument()
        End Sub
#Region "IDocumentContent"
        Sub CloseDocument()
            DirectCast(Me, IDocumentContent).DocumentOwner.Close(Me)
        End Sub
        ReadOnly Property IDocumentContent_Title() As Object Implements IDocumentContent.Title
            Get
                Return String.Empty
            End Get
        End Property
        Property IDocumentContent_DocumentOwner() As IDocumentOwner Implements IDocumentContent.DocumentOwner
        Sub IDocumentContent_OnClose(ByVal e As CancelEventArgs) Implements IDocumentContent.OnClose
        End Sub
        Sub IDocumentContent_OnDestroy() Implements IDocumentContent.OnDestroy
        End Sub
#End Region ' IDocumentContent
    End Class
End Namespace

File: CS/ViewModels/SignInViewModel.cs
namespace DXHtmlMessengerSample.ViewModels {
    using System.ComponentModel;
    using System.ComponentModel.DataAnnotations;
    using DevExpress.Mvvm;
    using DevExpress.Mvvm.POCO;

    public class SignInViewModel : IDocumentContent {
        public static SignInViewModel Create(string userName, string salt) {
            return ViewModelSource.Create(() => new SignInViewModel(userName, salt));
        }
        //
        readonly string salt;
        protected SignInViewModel(string userName, string salt) {
            this.UserName = userName;
            this.salt = salt;
        }
        public string UserName {
            get;
            private set;
        }
        [DataType(DataType.Password)]
        public virtual string Password {
            get;
            set;
        }
        public string AccessToken {
            get;
            private set;
        }
        public void ShowDialog() {
            var flyout = this.GetService<IDocumentManagerService>("Flyout");
            var document = flyout.CreateDocument(this);
            document.Show();
        }
        public void SignInViaSocialNetwork() {
            AccessToken = DevExpress.DevAV.Chat.DevAVEmpployeesInMemoryServer.GetPasswordHash(string.Empty, salt);
            CloseDocument();
        }
        public void SignIn() {
            AccessToken = DevExpress.DevAV.Chat.DevAVEmpployeesInMemoryServer.GetPasswordHash(Password, salt);
            CloseDocument();
        }
        #region IDocumentContent
        void CloseDocument() {
            ((IDocumentContent)this).DocumentOwner?.Close(this);
        }
        object IDocumentContent.Title {
            get { return string.Empty; }
        }
        IDocumentOwner IDocumentContent.DocumentOwner { get; set; }
        void IDocumentContent.OnClose(CancelEventArgs e) { }
        void IDocumentContent.OnDestroy() { }
        #endregion IDocumentContent
    }
}